### ISSUE TITLE
[token bill] Plan D: filter out more IRRELEVANT pages before getting into analyzer

### ISSUE THREAD

**Issue by shixuan** [shixuan/crawl-me-maybe#36]
Straightforward but complicated.

For past 5 runs:

run | found links | to ranker | ranker discarded | discard% | kept | analyzed | relevant | missed%
---  | ---      | ---| ---| ---| ---| ---| ---| ---
v2.6 strong conds  | 299      | 275    | 188       | 68% | 87   | 87     | 31       |64%    
 v2.6 split  | 289      | 274    | 141       | 51% | 133  | 133    | 37       | 72%   
 v2.5 fix seen | 298      | 274    | 158       | 58% | 116  | 116    | 38       | 67%    
 v2.5 LOW     | 304      | 280    | 172       | 61% | 108  | 108    | 41       | 62%    
 v2.5 MED     | 383      | 313    | 176       | 56% | 137  | 133    | 28       | 79%    
Total         |          | 1416   | 835       | 59% | 581  | 577    | 175      | 70%

**Comment by shixuan** [comment]
### Unsupervised

#### 1. Cosine to the goal
e5-large, query/passage prefixes, body truncated at 1600 characters.

run | AUC | mean, relevant | mean, irrelevant
-- | -- | -- | --
20260908_205341 | 0.800 | 0.821 | 0.805
20260908_180419 | 0.748 | 0.816 | 0.802
20260906_235310 | 0.760 | 0.815 | 0.802
20260905_212603 | 0.733 | 0.808 | 0.796
20260905_185808 | 0.707 | 0.805 | 0.792
seven runs pooled | 0.631 | — | —

The two means differ by 0.01–0.02, with the whole distribution packed into the band 0.79–0.87, and they overlap badly: the highest irrelevant scores 0.868, the lowest relevant 0.767. Any single cut takes from both sides.

Pooled AUC of 0.631 is lower than any individual run, which is the important part: the absolute scale drifts between tasks, so 0.81 here and 0.81 there do not mean the same thing. One fixed threshold that works across tasks was the entire point.

#### 2. One cut (Otsu, maximum between-class variance)

gap either side of the optimal cut:  +0.0009  +0.0008  +0.0004  +0.0015  +0.0015

Ten-thousandths. Not a gap, just a line drawn across a continuous distribution. It hardly matters where it goes, because there are not two clusters. Cutting without labels needs the data to fall into two lumps, and it is unimodal.

#### 3. A hand-built contrastive axis
Rather than one point, define a direction: sim(positive) − sim(negative). Still no labels.

pos: a post announcing a free gift, gift with purchase, or new merchandise giveaway at a merchant
neg: a post about a new product, a new flavour, a store opening, or general brand promotion

AUC 0.720–0.743 per run, 0.740 pooled — under plain cosine's 0.746. And those two sentences were written against the known confusion (new-flavour posts against giveaway posts), so this is the favourable case.

#### 4. Kernels and clustering

method | best component, AUC
-- | --
linear PCA | PC4 = 0.826
kernel PCA, cosine | PC4 = 0.840
kernel PCA, RBF γ=0.001 | 0.777
kernel PCA, RBF γ=0.01 | 0.658
kernel PCA, RBF γ=0.1 | 0.646

The higher the lift, the worse it gets (0.777 → 0.658 → 0.646). What a kernel amplifies is the dominant structure — topic, account, phrasing — and relevance is not that structure.

And that 0.840 on PC4 was picked using the labels: thirty components were computed and the one agreeing best with the labels was chosen. Unsupervised, nothing tells you to take the fourth; PC1–PC3 all carry more variance. Choosing correctly among thirty already requires the labels. Which is the point in miniature: relevance is a direction in the data, just not a prominent one, and unsupervised methods rank by prominence.

clustering | agreement with labels | ARI
-- | -- | --
KMeans k=2 | 0.654 | +0.076
spectral, RBF | 0.712 | +0.000
spectral, kNN graph | 0.695 | −0.019

ARI is zero throughout: the clusters have nothing to do with the relevant / irrelevant split. Agreement of 0.65–0.71 looks respectable only because 71% of pages are irrelevant — answering "irrelevant" every time scores 0.712, which is exactly what spectral RBF did.

**Comment by shixuan** [comment]
### Framing it as a general task

Right in principle, wrong in measurement.

"Does this passage answer this query" really is a general task: it is reranking, and there are pretrained models for it. A cross-encoder reads query and passage together and can attend across them, which is architecturally the right shape for "does this satisfy that". If a zero-shot reranker worked, the whole problem would be solved: no training, no cross-run state, transferable by construction.

model | pooled AUC
-- | --
ms-marco MiniLM-L-12 (English) | 0.661
jina-reranker-v2 multilingual (1.1GB) | 0.673
for comparison: cosine | 0.746

Both come in under cosine. Going multilingual moved 0.661 to 0.673, so the bottleneck is not language.

Why the framing fails. In the corpora rerankers are trained on, "answers" is very nearly "is about". The goal here is non-food gift-with-purchase promotions from merchants in the Greater Toronto Area, where "non-food", "in the GTA" and "a giveaway rather than a new product" are conditions, not topics. A post about a Toronto bubble tea shop's new flavour matches the topic closely and is exactly what must be refused. A reranker learns relevant; this needs satisfies. The two coincide in retrieval corpora and diverge systematically here.

**Comment by shixuan** [comment]
### Supervised

The only one that works, and the only one with a bill.

The labels need no annotation — analyses.classification is the label. The features are the body's e5 vector. The classifier is a linear SVM. All that is left is whether it generalises or memorises.

**The first numbers, and why they could not be trusted**
Leave-one-run-out gave AUC 0.914–0.988. Close to perfect, and therefore worth checking:

```
test pages whose exact text the training runs had already seen:
  20260908_205341: 74/98 = 76%
  20260908_180419: 72/93 = 77%
  20260906_235310: 80/125 = 64%
  20260905_212603: 85/119 = 71%
  20260905_185808: 86/152 = 57%
```

57%–77% of it was recall, not generalisation. The same Instagram posts are re-fetched days apart, and carousel entries repeat a caption word for word.

**With the leakage removed**

goal | test run | unseen n | AUC, all | AUC, unseen
-- | -- | -- | -- | --
giveaways | 20260908_205341 | 24 | 0.968 | 0.926
giveaways | 20260908_180419 | 21 | 0.988 | 0.978
giveaways | 20260906_235310 | 45 | 0.961 | 0.917
giveaways | 20260905_212603 | 34 | 0.952 | 0.896
giveaways | 20260905_185808 | 66 | 0.914 | 0.792
new drinks | 20260821_234041 | 25 | 0.918 | 0.895
new drinks | 20260825_211748 | 10 | 0.984 | 0.857
release notes | 20260821_233236 | 21 | 0.981 | 0.938
release notes | 20260822_080323 | 93 | 0.909 | 0.849

It replicates on three unrelated goals, including English technical content. But weight the two largest folds: n=93 gives 0.849 and n=66 gives 0.792. The honest floor is about 0.8, not 0.9+; the 0.978s and 1.000s come from folds of ten or twenty.

**The operating point, which is the number that decides**

AUC does not convert to savings. The question is how much of the reject pile disappears at a recall you can live with.

goal | test run | unseen n | irrelevant removed
-- | -- | -- | --
giveaways | 20260908_180419 | 21 | 93%
giveaways | 20260906_235310 | 45 | 87%
giveaways | 20260908_205341 | 24 | 78%
release notes | 20260821_233236 | 21 | 60%
new drinks | 20260821_234041 | 25 | 50%
release notes | 20260822_080323 | 93 | 29%
giveaways | 20260905_185808 | 66 | 27%
giveaways | 20260905_212603 | 34 | 21%

21% to 93%, a fourfold spread inside a single goal. And the pattern runs the wrong way: the three largest folds (93, 66, 45) gave 29%, 27% and 87% — the two biggest samples gave the two worst results, which makes the high scores on folds of twenty look like luck.

```
Taking the two largest folds:  removes 30% of the rejects, costs 5% of the results

A typical run of 100 pages (62 irrelevant, 38 relevant):
  today        62 × 1,699 + 38 × 4,700  =  284k
  SVM first    saves 62 × 30% × 1,699   =   32k  (11%)
               costs 38 × 5%            ≈  2 results
```
A result's marginal cost is 142k tokens. This trade sells something dearer than it buys.

**And the bill is not only those two results**
- Cross-run persistent state — the thing R7 warned about and C4 removed (domain_prior was "the only mutable global state breaking run isolation"). This time it is a file of weights, harder to point at than a dictionary when asking why this run came out the way it did.
- A model per goal, usable only after one full run. A task with a low hit rate (rentals: 2 relevant in 157) has too few positives to train on at all — and that is precisely where filtering is most wanted.
- Three new dependencies: fastembed, sklearn, and a 2.24GB local model.
- A feedback trap: pages the SVM filters never reach the analyzer and so never produce labels. The training set becomes only what the SVM already approves, the model drifts toward its own preferences, and nothing in the system can notice. That one is structural, and it matters more than the performance numbers.

**Comment by shixuan** [comment]
### The control group that turned up: the ranker already does most of it

Checking how much the ranker currently removes turned up a natural control in the stored runs: four of them dropped nothing at all, because of --recall or a zero threshold.


run | drop rate | hit rate among survivors
-- | -- | --
20260822_080323 | 0% | 22%
20260822_084349 | 0% | 37%
20260822_094648 | 0% | 39%
20260822_142247 | 68% | 97%
20260822_143922 | 66% | 92%
20260822_140043 | 51% | 80%

22–39% becomes 80–97%. The ranker is not only cutting volume, it is enriching — across 34 runs its drop rate is a median 61% (range 47%–79%).

Which explains why an embedding layer has so little left to do: it is handed a pool that has already been enriched once. The easy rejections are gone and what remains is expensive to judge and cheap to keep.

**Comment by shixuan** [comment]
### The accounting that came last, and what it changed

run | judged | relevant | total tokens | per result
-- | -- | -- | -- | --
20260902_191230 | 27 | 5 | 145,239 | 29,048
20260822_080323 | 118 | 25 | 449,940 | 17,998
… | — | — | — | —
20260822_140043 | 92 | 73 | 299,001 | 4,096
20260822_142247 | 66 | 59 | 241,395 | 4,091

```
30 runs  tokens per result: median 10,544   best 4,091   worst 29,048
at roughly $0.35/M:  a result $0.004     a run (median 338k) $0.12
```

Four tenths of a cent per result, twelve cents per run. Set against what this effort produced — two calls instead of one (0%), a conditional contract (0%), the admission gate (0.9%), prompt cache order (backwards), six embedding methods (none) — the time spent far exceeded the money available to save.

The real spread is 7×, and it is not about token efficiency:
```
best   20260822_142247   66 judged → 59 relevant (89%)   per result  4,091
worst  20260902_191230   27 judged →  5 relevant (19%)   per result 29,048
```

All of it is hit rate, and on this evidence hit rate follows how well the goal matches the seeds. The strong runs ask for what the seeds can supply (bubble tea shops' new drinks, from bubble tea accounts); the weak ones carry conditions narrower than anything the seeds hold, and the crawler hunts through pages that could never have satisfied them.

Saving tokens is bounded by the change in 10,544; hit rate is a factor of seven. This effort went into the shortest lever on the board, which is the most valuable line in this archive.

If this layer is proposed again, start here
- Unsupervised fails for lack of information, not for lack of quality. Relevance is a direction in the data (PC4 reaches 0.84), just not a prominent one, and unsupervised methods rank by prominence.
- The "general task" framing does not hold. A pretrained reranker learns topical relevance; this needs conditional satisfaction. They coincide in retrieval corpora and diverge here.
- Check for text leakage before believing any leave-one-run-out on repeated seeds. It was 57%–77% here, and skipping the check inflates the result by a full grade.
- AUC is not the saving. An AUC of 0.85 removed 30% of rejects at the 95%-recall operating point, and that is the number money is spent against.
- Price a result first. That figure belongs at the start of an optimisation, not after abandoning it.
